generic/piranha-tree-sitter/src/piranha.rs

```diff
@@ -0,0 +1,504 @@
+/*
+Copyright (c) 2022 Uber Technologies, Inc.
+
+ <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ except in compliance with the License. You may obtain a copy of the License at
+ <p>http://www.apache.org/licenses/LICENSE-2.0
+
+ <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ express or implied. See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+//! This module contains the core logic for Piranha.
+use crate::config::{
+    command_line_arguments::PiranhaArguments, Rule, RuleStore, CLASS, GLOBAL, METHOD, PARENT,
+};
+use crate::tree_sitter::{PiranhaRuleMatcher, TreeSitterHelpers};
+use crate::utilities::read_file;
+use colored::Colorize;
+use itertools::Itertools;
+use jwalk::WalkDir;
+use regex::Regex;
+use std::{collections::HashMap, fs, path::PathBuf};
+use tree_sitter::{InputEdit, Node, Parser, Point, Range, Tree};
+
+// Maintains the state of Piranha and the updated content of files in the source code.
+pub struct FlagCleaner {
+    // Maintains Piranha's state
+    rule_store: RuleStore,
+    // Path to source code folder
+    path_to_codebase: String,
+    // Files updated by Piranha.
+    pub relevant_files: HashMap<PathBuf, SourceCodeUnit>,
+}
+
+impl FlagCleaner {
+    // Performs cleanup related to stale flags
+    pub fn cleanup(&mut self) {
+        let mut parser = Parser::new();
+        parser
+            .set_language(self.rule_store.piranha_args.language)
+            .expect("Could not set the language for the parser.");
+
+        loop {
+            let current_rules = self.rule_store.get_global_rules();
+
+            println!("Number of global rules {}", current_rules.len());
+
+            let files_containing_pattern = self.get_relevant_files();
+
+            #[rustfmt::skip]
+            println!("{}", format!("Will parse and analyze {} files.", files_containing_pattern.len()).green());
+
+            for (path, content) in files_containing_pattern {
+                self.relevant_files
```

Comment:
comment on what this does. in general add comments throughout the file to help the reader (especially, given this is in Rust).